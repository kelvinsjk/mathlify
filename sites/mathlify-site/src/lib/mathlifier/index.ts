import { renderHTML } from 'djot-temml';

/**
 * generates djot string,
 * using JavaScript's template literal syntax to handle mathematical expressions (both static and dynamic)
 *
 * math: starts with ${}, terminates with \n,
 * display: starts with $${}, terminates with \n\n,
 * display env: starts with $${'align'}, etc.
 * text: starts with @${}, terminates immediately
 */
export function mathlifyDj(strings: TemplateStringsArray, ...values: unknown[]): string {
	return mathlifyGen(modules)(strings, ...values);
}

export function mathlify(strings: TemplateStringsArray, ...values: unknown[]): string {
	return renderHTML(mathlifyDj(strings, ...values));
}

export function mathlifyQED(strings: TemplateStringsArray, ...values: unknown[]): string {
	return renderHTML(
		mathlifyGen(modules, { postContent: { mode: 'auto-math', content: '\\; \\blacksquare' } })(
			strings,
			...values,
		),
	);
}

const modules: Modules = {
	math: (x: string) => `$\`${x}\``,
	display,
	mathEnvs: {
		equation: (x: string) => display(insertEnv('equation', x)),
		align: (x: string) => display(insertEnv('align', x)),
		gather: (x: string) => display(insertEnv('gather', x)),
		alignat: (x: string) => display(insertEnv('alignat', x)),
		equationStar: (x: string) => display(insertEnv('equation*', x)),
		alignStar: (x: string) => display(insertEnv('align*', x)),
		gatherStar: (x: string) => display(insertEnv('gather*', x)),
		alignatStar: (x: string) => display(insertEnv('alignat*', x)),
	},
};

function insertEnv(env: string, content: string, args = ''): string {
	if (env === 'alignat' || env === 'alignat*') {
		let firstLine = content.split('\n')[0];
		if (!firstLine) {
			firstLine = content.split('\n')[1];
		}
		const twoN = firstLine.split('&').length;
		if (twoN % 2 !== 0)
			console.warn(`alignat should have an even number of columns, but got ${twoN} instead`);
		args = `{${Math.floor(twoN / 2)}}`;
	}
	// todo: handle this with regex
	while (content.endsWith('\n') || content.endsWith('\t') || content.endsWith(' ')) {
		content = content.slice(0, content.length - 1);
	}
	return `\\begin{${env}}${args}${content}\n\\end{${env}}`;
}

function display(x: string): string {
	return `$$\`${x}\``;
}

enum Modes {
	text, // 0
	math, // 1
	display, // 2
	mathEnv, // 3
}

export interface Modules {
	math: (x: string) => string;
	display: (x: string) => string;
	mathEnvs?: {
		equation?: (x: string) => string;
		align?: (x: string) => string;
		gather?: (x: string) => string;
		alignat?: (x: string) => string;
		equationStar?: (x: string) => string;
		alignStar?: (x: string) => string;
		gatherStar?: (x: string) => string;
		alignatStar?: (x: string) => string;
	};
}

enum MathEnvs {
	equation,
	equationStar,
	align,
	alignStar,
	gather,
	gatherStar,
	alignat,
	alignatStar,
}

type postContentModes = 'text' | 'math' | 'display' | 'auto-math';

const mathEnvsKeys = [
	'equation',
	'equation*',
	'align',
	'align*',
	'gather',
	'gather*',
	'alignat',
	'alignat*',
] as const;
function isMathEnv(x: string): x is (typeof mathEnvsKeys)[number] {
	return (mathEnvsKeys as readonly string[]).includes(x);
}

const envToMode: Record<(typeof mathEnvsKeys)[number], MathEnvs> = {
	equation: MathEnvs.equation,
	'equation*': MathEnvs.equationStar,
	align: MathEnvs.align,
	'align*': MathEnvs.alignStar,
	gather: MathEnvs.gather,
	'gather*': MathEnvs.gatherStar,
	alignat: MathEnvs.alignat, // triggered by $${'alignat'}{x}
	'alignat*': MathEnvs.alignatStar, // triggered by $${'alignat'}{x}
};

interface MathEnvOptions {
	mathEnv: MathEnvs;
	eqnColumns?: number;
}

// TODO: alignat cols

export function mathlifyGen(
	modules: Modules,
	options?: {
		postContent?: { mode?: postContentModes; content: string };
	},
): (strings: TemplateStringsArray, ...values: unknown[]) => string {
	return (strings: TemplateStringsArray, ...values: unknown[]) => {
		let finalOutput = '';
		let accumulatedEnvOutput = '';
		let mode: Modes = Modes.text;
		let mathEnvOptions: MathEnvOptions = { mathEnv: MathEnvs.equation };
		strings.forEach((str, i) => {
			let nextVal = values[i];
			if (
				nextVal === undefined ||
				nextVal === null ||
				(typeof nextVal === 'object' && Object.keys(nextVal).length === 0)
			)
				nextVal = '';
			if (mode === Modes.text) {
				if (str.endsWith('@')) {
					// continues text mode
					finalOutput += `${str.slice(0, str.length - 1)}${nextVal}`;
				} else {
					// starts new text environment
					[mode, accumulatedEnvOutput, finalOutput, mathEnvOptions] = startNewEnv(
						str,
						'',
						nextVal,
						finalOutput,
					);
				}
			} else if (mode === Modes.math) {
				// checks for \n or \r\n non-greedily
				const regex = /([^]*?)(\r?\n)([^]*)/;
				const match = str.match(regex);
				if (match) {
					// end math mode
					const [, before, newline, after] = match;
					accumulatedEnvOutput += `${before}`;
					if (accumulatedEnvOutput) {
						finalOutput += modules.math(accumulatedEnvOutput);
					}
					[mode, accumulatedEnvOutput, finalOutput, mathEnvOptions] = startNewEnv(
						after,
						newline,
						nextVal,
						finalOutput,
					);
				} else {
					// continue math mode
					if (str.endsWith(`@`)) {
						// TODO: check this
						accumulatedEnvOutput += `${str.slice(0, str.length - 1)}`;
					} else {
						accumulatedEnvOutput += `${str}${nextVal}`;
					}
					// if (i === strings.length - 1) {
					// 	// final string
					// 	finalOutput += modules.math(accumulatedEnvOutput);
					// }
				}
			} else if (mode === Modes.display) {
				// checks for \n\n or \r\n\r\n within str. group into (#1)(#2) where #2 = \n or \r\n and it is done non-greedily
				const regex = /([^]*?)(\r?\n[ \t]*\r?\n)([^]*)/;
				const match = str.match(regex);
				if (match) {
					// end display mode
					const [, before, newline, after] = match;
					accumulatedEnvOutput += before;
					if (accumulatedEnvOutput) {
						finalOutput += modules.display(accumulatedEnvOutput);
					}
					[mode, accumulatedEnvOutput, finalOutput, mathEnvOptions] = startNewEnv(
						after,
						newline,
						nextVal,
						finalOutput,
					);
				} else {
					// continue display mode
					if (str.endsWith(`@`)) {
						// TODO: check this
						accumulatedEnvOutput += `${str.slice(0, str.length - 1)}`;
					} else {
						accumulatedEnvOutput += `${str}${nextVal}`;
					}
					// if (i === strings.length - 1) {
					// 	// final string
					// 	finalOutput += modules.display(accumulatedEnvOutput);
					// }
				}
			} else if (mode === Modes.mathEnv) {
				// checks for \n\n or \r\n\r\n within str. group into (#1)(#2) where #2 = \n or \r\n and it is done non-greedily
				const regex = /([^]*?)(\r?\n[ \t]*\r?\n)([^]*)/;
				const match = str.match(regex);
				if (match) {
					// end mathEnv mode
					const [, before, newline, after] = match;
					accumulatedEnvOutput += before;
					if (accumulatedEnvOutput) {
						finalOutput += insertMathEnv(mathEnvOptions, accumulatedEnvOutput, modules);
					}
					[mode, accumulatedEnvOutput, finalOutput, mathEnvOptions] = startNewEnv(
						after,
						newline,
						nextVal,
						finalOutput,
					);
				} else {
					// continue mathEnv mode
					if (str.endsWith(`@`)) {
						// TODO: check this.
						accumulatedEnvOutput += `${str.slice(0, str.length - 1)}`;
					} else {
						accumulatedEnvOutput += `${str}${nextVal}`;
					}
					// if (i === strings.length - 1) {
					// 	// final string
					// 	finalOutput += insertMathEnv(mathEnvOptions, accumulatedEnvOutput, modules);
					// }
				}
			}
		});
		// handle accumulated value and post content
		if (accumulatedEnvOutput) {
			const currentMode = mode as Modes;
			if (options?.postContent) {
				const finalMode = options.postContent.mode ?? 'auto-math';
				const content = options.postContent.content;
				if (currentMode === Modes.text && finalMode === 'text') {
					finalOutput += accumulatedEnvOutput + content;
				} else if (
					currentMode === Modes.math &&
					(finalMode === 'math' || finalMode === 'auto-math')
				) {
					finalOutput += modules.math(accumulatedEnvOutput + content);
				} else if (
					currentMode === Modes.display &&
					(finalMode === 'display' || finalMode === 'auto-math')
				) {
					finalOutput += modules.display(accumulatedEnvOutput + content);
				} else if (currentMode === Modes.mathEnv && finalMode === 'auto-math') {
					finalOutput += insertMathEnv(mathEnvOptions, accumulatedEnvOutput + content, modules);
				} else {
					// new env. append final value first
					if (currentMode === Modes.text) {
						finalOutput += accumulatedEnvOutput;
					} else if (currentMode === Modes.math) {
						finalOutput += modules.math(accumulatedEnvOutput);
					} else if (currentMode === Modes.display) {
						finalOutput += modules.display(accumulatedEnvOutput);
					} else {
						finalOutput += insertMathEnv(mathEnvOptions, accumulatedEnvOutput, modules);
					}
					// handle post content
					if (finalMode === 'text') {
						finalOutput += content;
					} else if (finalMode === 'math' || finalMode === 'auto-math') {
						finalOutput += modules.math(content);
					} else if (finalMode === 'display') {
						finalOutput += modules.display(content);
					}
				}
			} else {
				if (currentMode === Modes.math) {
					finalOutput += modules.math(accumulatedEnvOutput);
				} else if (currentMode === Modes.display) {
					finalOutput += modules.display(accumulatedEnvOutput);
				} else {
					finalOutput += insertMathEnv(mathEnvOptions, accumulatedEnvOutput, modules);
				}
			}
		} else if (options?.postContent) {
			const { mode, content } = options.postContent;
			if (mode === 'text') {
				finalOutput += content;
			} else if (mode === 'math' || mode === 'auto-math') {
				finalOutput += modules.math(content);
			} else if (mode === 'display') {
				finalOutput += modules.display(content);
			}
		}
		return finalOutput;
	};
}

// returns [Mode, acc, curr, {mathEnv, eqnColumns?}]
function startNewEnv(
	after: string,
	newline: string,
	nextVal: unknown,
	curr: string,
): [Modes, string, string, MathEnvOptions] {
	const defaultMathEnvOptions = { mathEnv: MathEnvs.equation };
	let mathEnvOptions: MathEnvOptions = defaultMathEnvOptions;
	if (after.endsWith('$')) {
		let x = `${nextVal}`;
		let mode = Modes.display;
		if (isMathEnv(x)) {
			// new math env
			mode = Modes.mathEnv;
			mathEnvOptions = { mathEnv: envToMode[x] };
			//TODO: handle alignat cols
			x = '';
		}
		// new display mode
		return [mode, x, curr + `${newline}${after.slice(0, after.length - 1)}`, mathEnvOptions];
	} else if (after.endsWith('@')) {
		// new text mode
		return [
			Modes.text,
			'',
			curr + `${newline}${after.slice(0, after.length - 1)}${nextVal}`,
			defaultMathEnvOptions,
		];
	} else {
		// new math mode
		return [Modes.math, `${nextVal}`, curr + newline + after, defaultMathEnvOptions];
	}
}

function insertMathEnv(mathEnvOptions: MathEnvOptions, acc: string, modules: Modules): string {
	const mathEnv = mathEnvOptions.mathEnv;
	if (mathEnv === MathEnvs.equation) {
		return modules.mathEnvs?.equation ? modules.mathEnvs.equation(acc) : acc;
	} else if (mathEnv === MathEnvs.equationStar) {
		return modules.mathEnvs?.equationStar ? modules.mathEnvs.equationStar(acc) : acc;
	} else if (mathEnv === MathEnvs.align) {
		return modules.mathEnvs?.align ? modules.mathEnvs.align(acc) : acc;
	} else if (mathEnv === MathEnvs.alignStar) {
		return modules.mathEnvs?.alignStar ? modules.mathEnvs.alignStar(acc) : acc;
	} else if (mathEnv === MathEnvs.gather) {
		return modules.mathEnvs?.gather ? modules.mathEnvs.gather(acc) : acc;
	} else if (mathEnv === MathEnvs.gatherStar) {
		return modules.mathEnvs?.gatherStar ? modules.mathEnvs.gatherStar(acc) : acc;
	} else if (mathEnv === MathEnvs.alignat) {
		return modules.mathEnvs?.alignat ? modules.mathEnvs.alignat(acc) : acc;
	} else if (mathEnv === MathEnvs.alignatStar) {
		return modules.mathEnvs?.alignatStar ? modules.mathEnvs.alignatStar(acc) : acc;
	}
	console.warn(`mathEnv ${mathEnv} not found`);
	return acc;
}
