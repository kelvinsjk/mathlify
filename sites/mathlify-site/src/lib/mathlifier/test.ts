import { mathlify } from './';

const str = mathlify`This is inline math: ${'x+2'}xyz${'k'}
and this is text.

This is display math: $${'x+2'}xyz
abcd
${'k'}

This is aligned:
$${'align'}
  x+2 &= ${'k'} \\\\
  y &= ${'z'}`;

console.log(str);
