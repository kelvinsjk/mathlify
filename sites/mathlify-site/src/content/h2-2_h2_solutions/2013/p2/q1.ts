import { Answer } from '$lib/classes/answer';
import { fnTerm, Polynomial, quotient } from 'mathlify';
import { generateDomain } from '$content/h2-1_h2_learn/01_functions/02_inverse/02_domain/02_practice';
import { compositeExists } from '$content/h2-1_h2_learn/01_functions/03_composite/01_existence/02_practice';
import { compositeFormula } from '$content/h2-1_h2_learn/01_functions/03_composite/02_formula/02_practice';
import { EquationWorking } from 'mathlify/working';
import { mathlifierDj as mathlifier } from 'mathlifier';
import { Interval } from '$content/h2-1_h2_learn/01_functions/_intervals';
import { QED } from '$lib/utils/typesetting/utils';

export const answer = new Answer();

// ! Complete
// f(x) = (2+x)/(1-x)
// g(x) = 1-2x

const [a, b, c] = [-1, 1, 2];
const ascending = true;
const f = quotient(new Polynomial([c, b], { ascending }), new Polynomial([1, -1], { ascending }));
const g = new Polynomial([1, -2], { ascending });
const rhs = 5;
const fState = {
	fnType: 'improper',
	a,
	b,
	c,
	restriction: false,
	unknownConstants: false
} as const;
const gState = {
	fnType: 'linear',
	a: -2,
	b: 1,
	c: 0,
	restriction: false,
	unknownConstants: false
} as const;
const compositeStateFg: Parameters<typeof compositeExists>[0] = {
	f: fState,
	g: gState,
	fg: true
};

// a
{
	const { ans, soln } = compositeExists(compositeStateFg, [f, g]);
	answer.addPart(ans, soln);
}
// b
{
	const {
		ans: ans1,
		soln: soln1,
		exp: gf
	} = compositeFormula([f, generateDomain(fState), g, [Interval.ALL_REAL]], false, false, {
		noDomain: true,
		ansInline: true,
		combineFraction: true,
		QED: true
	});

	const gfx = fnTerm('gf', 'x');
	const working = new EquationWorking(gfx, rhs, { aligned: true });
	working.subIn({ gf }).crossMultiply({ hide: true }).expand();
	const { root } = working.solve.linear('x');
	const soln =
		soln1 +
		mathlifier`$${'align*'} \\text{Let } \\quad x &= (gf)^{-1}(${rhs})  \\\\ ${working}
	
	$${{}} (gf)^{-1} (${rhs}) = ${root} \\; ${QED}`;
	const ans = ans1 + '\\\n' + mathlifier`${{}} (gf)^{-1} (${rhs}) = ${root}.`;
	answer.addPart(ans, soln);
}
