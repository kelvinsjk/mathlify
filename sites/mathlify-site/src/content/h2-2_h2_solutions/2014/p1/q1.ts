import { Answer } from '$lib/classes/answer';
import { mathlifierDj as mathlifier } from 'mathlifier';
import { Expression, Polynomial } from 'mathlify';
import { fractionalInverseWorking } from '$content/h2-1_h2_learn/01_functions/02_inverse/03_formula/02_practice';
import { compositeFormula } from '$content/h2-1_h2_learn/01_functions/03_composite/02_formula/02_practice';
import { QED } from '$lib/utils/typesetting/utils';

export const answer = new Answer();

// ! Complete
// 1/(1-x)

const f = new Expression([1, '/', new Polynomial([1, -1], { ascending: true })]);

// a
{
	const { soln: f2Working, exp: f2 } = compositeFormula([f, [], f, []], true, false, {
		gName: 'f',
		noDomain: true
	});
	const { working: fInvWorking } = fractionalInverseWorking(f, false, { reportInverse: true });
	const soln = f2Working + fInvWorking + mathlifier`Hence ${'f^2(x) = f^{-1}(x)'} \\; ${QED}`;
	const ans = mathlifier`${{}}f^2(x)=f^{-1}(x) = ${f2}.`;
	answer.addPart(ans, soln);
}
// b
{
	const soln = mathlifier`$${'align*'} f^3(x) &= f^2f(x) \\\\ &= f^{-1}f(x) \\quad \\text{since } f^2(x)=f^{-1}(x) \\\\ &= x \\; ${QED}`;
	const ans = mathlifier`${{}}f^3(x)=x.`;
	answer.addPart(ans, soln);
}
