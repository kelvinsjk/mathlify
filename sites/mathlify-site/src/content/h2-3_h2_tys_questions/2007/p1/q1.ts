/* The question here is the copyright of 
- Singapore Examinations and Assessment Board
- Cambridge Assessment International Education
- Ministry of Education, Singapore
- Government of Singapore

We provide this question on Mathlify under fair use for archival purposes.
These are questions more than 10 years old and are difficult to obtain
commercially.

For the newer questions, please purchase a copy of the
"ten year series" from authorized publishers and distributors.
*/

import { Polynomial, quotient, sum } from 'mathlify';
import { combineFraction } from 'mathlify/algebra';
import { Question } from '$lib/classes/question';
import { mathlifierDj as mathlifier } from 'mathlifier';

export const question = new Question();

// (2x^2 - x - 19) / (x^2 + 3x + 2)
// (x^2 - 4x - 21) / (x^2 + 3x + 2)

const num = new Polynomial([2, -1, -19]);
const den = new Polynomial([1, 3, 2]);
const frac = quotient(num, den);
const secondTerm = 1;
const lhs = sum(frac, -secondTerm);
const rhs = combineFraction(lhs);

// a
{
	question.addPart(
		mathlifier`Show that

$${lhs} = ${rhs}.`,
		{ marks: 1 }
	);
}
// b
{
	question.addPart(
		mathlifier`Hence, without using a calculator, solve the inequality

$${frac} > ${secondTerm}.`,
		{ marks: 4 }
	);
}
