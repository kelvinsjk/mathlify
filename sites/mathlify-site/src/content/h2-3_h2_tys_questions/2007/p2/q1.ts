/* The question here is the copyright of 
- Singapore Examinations and Assessment Board
- Cambridge Assessment International Education
- Ministry of Education, Singapore
- Government of Singapore

We provide this question on Mathlify under fair use for archival purposes.
These are questions more than 10 years old and are difficult to obtain
commercially.

For the newer questions, please purchase a copy of the
"ten year series" from authorized publishers and distributors.
*/

import { Question } from '$lib/classes/question';
import { mathlifierDj as mathlifier } from 'mathlifier';

export const question = new Question();

const matrix = [
	[1.15, 0.6, 0.55, 8.28],
	[1.2, 0.45, 0.3, 6.84],
	[2.15, 0.9, 0.65, 13.05]
];
const leeLian = [1.3, 0.25, 0.5];
const rows = ['| :-- | :---: | :---: | :---: | :---: |', '| | Suresh | Fandi | Cindy | Lee Lian |'];
const headings = ['Pineapples (kg)', 'Mangoes (kg)', 'Lychees (kg)', 'Total amount paid in \\$'];
for (let i = 0; i < 4; i++) {
	rows.push(
		`| ${headings[i]} | ${matrix[0][i].toFixed(2)} | ${matrix[1][i].toFixed(2)} | ${matrix[2][i].toFixed(2)} | ${leeLian[i]?.toFixed(2) ?? ''} |`
	);
}
const table = rows.join('\n');

question.addBody(
	mathlifier`Four friends buy three different kinds of fruit in the market. When they get home they cannot remember
the individual prices per kilogram, but three of them can remember the total amount that they each
paid.

The weights of fruit and the total amounts paid are shown in the following table.

@${table}

Assuming that, for each variety of fruit, the price per kilogram paid by each of the friends is the same,
calculate the total amount that Lee Lian paid.`,
	{ marks: 6 }
);
