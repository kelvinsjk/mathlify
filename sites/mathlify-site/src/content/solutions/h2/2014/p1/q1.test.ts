import { expect, it } from 'vitest';
import { answer } from './q1.soln';
import { mathlify } from '$lib/mathlifier';

const ans = answer.answer;
const actual: typeof ans = {
	parts: [
		{
			body: mathlify`${{}} f^2(x) = f^{-1}(x) = \\frac{x-1}{x}.`,
		},
		{ body: mathlify`${{}} f^3(x)=x.` },
	],
};

it('2014p1q1', () => {
	expect(answer).toMatchSnapshot();
	expect(ans).toEqual(actual);
});
