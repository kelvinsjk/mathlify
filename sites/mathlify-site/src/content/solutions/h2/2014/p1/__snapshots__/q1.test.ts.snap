// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`2014p1q1 1`] = `
Answer {
  "answer": {
    "parts": [
      {
        "body": "<p><math><mrow><msup><mi>f</mi><mn>2</mn></msup><mo form="prefix" stretchy="false">(</mo><mi>x</mi><mo form="postfix" stretchy="false">)</mo><mo>=</mo></mrow><mrow><msup><mi>f</mi><mrow><mo form="prefix" stretchy="false" lspace="0em" rspace="0em">−</mo><mn>1</mn></mrow></msup><mo form="prefix" stretchy="false">(</mo><mi>x</mi><mo form="postfix" stretchy="false">)</mo><mo>=</mo></mrow><mrow><mo form="prefix" stretchy="false">−</mo></mrow><mrow><mfrac><mrow><mn>1</mn><mo>−</mo><mi>x</mi></mrow><mi>x</mi></mfrac><mi>.</mi></mrow></math></p>
",
      },
      {
        "body": "<p><math><mrow><msup><mi>f</mi><mn>3</mn></msup><mo form="prefix" stretchy="false">(</mo><mi>x</mi><mo form="postfix" stretchy="false">)</mo><mo>=</mo></mrow><mrow><mi>x</mi><mi>.</mi></mrow></math></p>
",
      },
    ],
  },
  "solution": {
    "parts": [
      {
        "body": "<p><math display="block" class="tml-display" style="display:block math;"><mtable displaystyle="true" columnalign="right left"><mtr><mtd class="tml-right" style="padding:0.7ex 0em 0.7ex 0em;"><mrow><msup><mi>f</mi><mn>2</mn></msup><mo form="prefix" stretchy="false">(</mo><mi>x</mi><mo form="postfix" stretchy="false">)</mo></mrow></mtd><mtd class="tml-left" style="padding:0.7ex 0em 0.7ex 0em;"><mrow><mo>=</mo><mi>f</mi><mrow><mo fence="true" form="prefix">(</mo><mfrac><mn>1</mn><mrow><mn>1</mn><mo>−</mo><mi>x</mi></mrow></mfrac><mo fence="true" form="postfix">)</mo></mrow></mrow></mtd></mtr><mtr><mtd class="tml-right" style="padding:0.7ex 0em 0.7ex 0em;"><mrow></mrow></mtd><mtd class="tml-left" style="padding:0.7ex 0em 0.7ex 0em;"><mrow><mo>=</mo><mfrac><mn>1</mn><mrow><mn>1</mn><mo>−</mo><mfrac><mn>1</mn><mrow><mn>1</mn><mo>−</mo><mi>x</mi></mrow></mfrac></mrow></mfrac></mrow></mtd></mtr><mtr><mtd class="tml-right" style="padding:0.7ex 0em 0.7ex 0em;"><mrow></mrow></mtd><mtd class="tml-left" style="padding:0.7ex 0em 0.7ex 0em;"><mrow><mo>=</mo><mfrac><mn>1</mn><mrow><mo lspace="0em" rspace="0em">−</mo><mfrac><mi>x</mi><mrow><mn>1</mn><mo>−</mo><mi>x</mi></mrow></mfrac></mrow></mfrac></mrow></mtd></mtr><mtr><mtd class="tml-right" style="padding:0.7ex 0em 0.7ex 0em;"><mrow></mrow></mtd><mtd class="tml-left" style="padding:0.7ex 0em 0.7ex 0em;"><mrow><mo>=</mo><mo form="prefix" stretchy="false">−</mo><mfrac><mrow><mn>1</mn><mo>−</mo><mi>x</mi></mrow><mi>x</mi></mfrac></mrow></mtd></mtr></mtable></math></p>
<p><math display="block" class="tml-display" style="display:block math;"><mtable displaystyle="true"><mtr><mtd style="padding:0.7ex 0em 0.7ex 0em;"><mrow><mtext>Let </mtext><mi>y</mi><mo>=</mo><mfrac><mn>1</mn><mrow><mn>1</mn><mo>−</mo><mi>x</mi></mrow></mfrac></mrow></mtd></mtr><mtr><mtd style="padding:0.7ex 0em 0.7ex 0em;"><mrow><mi>y</mi><mrow><mo fence="true" form="prefix">(</mo><mn>1</mn><mo>−</mo><mi>x</mi><mo fence="true" form="postfix">)</mo></mrow><mo>=</mo><mn>1</mn></mrow></mtd></mtr><mtr><mtd style="padding:0.7ex 0em 0.7ex 0em;"><mrow><mi>y</mi><mo>−</mo><mi>y</mi><mi>x</mi><mo>=</mo><mn>1</mn></mrow></mtd></mtr><mtr><mtd style="padding:0.7ex 0em 0.7ex 0em;"><mrow><mo>−</mo><mi>y</mi><mi>x</mi><mo>=</mo><mn>1</mn><mo>−</mo><mi>y</mi></mrow></mtd></mtr><mtr><mtd style="padding:0.7ex 0em 0.7ex 0em;"><mrow><mi>x</mi><mo>=</mo><mo form="prefix" stretchy="false">−</mo><mfrac><mrow><mn>1</mn><mo>−</mo><mi>y</mi></mrow><mi>y</mi></mfrac></mrow></mtd></mtr><mtr><mtd style="padding:0.7ex 0em 0.7ex 0em;"><mrow><msup><mi>f</mi><mrow><mo lspace="0em" rspace="0em">−</mo><mn>1</mn></mrow></msup><mo form="prefix" stretchy="false">(</mo><mi>x</mi><mo form="postfix" stretchy="false">)</mo><mo>=</mo><mo form="prefix" stretchy="false">−</mo><mfrac><mrow><mn>1</mn><mo>−</mo><mi>x</mi></mrow><mi>x</mi></mfrac></mrow></mtd></mtr></mtable></math></p>
<p>Hence <math><mrow><msup><mi>f</mi><mn>2</mn></msup><mo form="prefix" stretchy="false">(</mo><mi>x</mi><mo form="postfix" stretchy="false">)</mo><mo>=</mo></mrow><mrow><msup><mi>f</mi><mrow><mo form="prefix" stretchy="false" lspace="0em" rspace="0em">−</mo><mn>1</mn></mrow></msup><mo form="prefix" stretchy="false">(</mo><mi>x</mi><mo form="postfix" stretchy="false">)</mo><mspace width="0.2778em"></mspace><mi>■</mi></mrow></math></p>
",
      },
      {
        "body": "<p><math display="block" class="tml-display" style="display:block math;"><mtable displaystyle="true" columnalign="right left"><mtr><mtd class="tml-right" style="padding:0.7ex 0em 0.7ex 0em;"><mrow><msup><mi>f</mi><mn>3</mn></msup><mo form="prefix" stretchy="false">(</mo><mi>x</mi><mo form="postfix" stretchy="false">)</mo></mrow></mtd><mtd class="tml-left" style="padding:0.7ex 0em 0.7ex 0em;"><mrow><mo>=</mo><msup><mi>f</mi><mn>2</mn></msup><mi>f</mi><mo form="prefix" stretchy="false">(</mo><mi>x</mi><mo form="postfix" stretchy="false">)</mo></mrow></mtd></mtr><mtr><mtd class="tml-right" style="padding:0.7ex 0em 0.7ex 0em;"><mrow></mrow></mtd><mtd class="tml-left" style="padding:0.7ex 0em 0.7ex 0em;"><mrow><mo>=</mo><msup><mi>f</mi><mrow><mo form="prefix" stretchy="false" lspace="0em" rspace="0em">−</mo><mn>1</mn></mrow></msup><mi>f</mi><mo form="prefix" stretchy="false">(</mo><mi>x</mi><mo form="postfix" stretchy="false">)</mo><mspace width="1em"></mspace><mtext>since </mtext><msup><mi>f</mi><mn>2</mn></msup><mo form="prefix" stretchy="false">(</mo><mi>x</mi><mo form="postfix" stretchy="false">)</mo><mo>=</mo><msup><mi>f</mi><mrow><mo form="prefix" stretchy="false" lspace="0em" rspace="0em">−</mo><mn>1</mn></mrow></msup><mo form="prefix" stretchy="false">(</mo><mi>x</mi><mo form="postfix" stretchy="false">)</mo></mrow></mtd></mtr><mtr><mtd class="tml-right" style="padding:0.7ex 0em 0.7ex 0em;"><mrow></mrow></mtd><mtd class="tml-left" style="padding:0.7ex 0em 0.7ex 0em;"><mrow><mo>=</mo><mi>x</mi><mspace width="0.2778em"></mspace><mi>■</mi></mrow></mtd></mtr></mtable></math></p>
",
      },
    ],
  },
}
`;
